### PR TITLE
Separate projection bounds and predicates

### PR DESCRIPTION
**Pull request by matthewjasper**
Follow up to #72788.

- Rename `projection_predicates` to `item_bounds`
- Separate bounds on associated types (the things after the `:` in `type X: ...`) and opaque types (the things after `impl`)  from predicates.
- Projection candidates now have the correct nested obligations
- Trait object candidates now check that the associated types on the trait object satisfy their bounds as nested obligations
- Type alias impl trait types are now checked (#73035)
- `feature(generic_associated_types)` no longer changes how we handle bounds (#73816)

Opening for a perf and crater runs.

r? @nikomatsakis

### CODE REVIEW

**Comment by matthewjasper** [comment]
@craterbot check

**Comment by matthewjasper** [comment]
Actually, it looks like some of the of the perf creates failed to build, I'll cancel if that's the case locally.

**Comment by matthewjasper** [comment]
@craterbot cancel

**Comment by matthewjasper** [comment]
@craterbot abort

File: src/librustc_typeck/check/mod.rs
Comment (nikomatsakis) [inline src/librustc_typeck/check/mod.rs:1906]: Can you elaborate a bit on the motivation here? Is this just a redundant check? i.e., each place that specifies the type of the opaque type is also doing these checks, right?

File: src/librustc_trait_selection/opaque_types.rs
Comment (nikomatsakis) [inline src/librustc_trait_selection/opaque_types.rs]: heh I've been meaning to change that "cause" info for years...

File: src/test/incremental/issue-54242.rs
Comment (nikomatsakis) [inline src/test/incremental/issue-54242.rs:3]: Uh..?

**Review by nikomatsakis — COMMENTED** [review]
Generally 👍, left some comments/thoughts

File: src/test/incremental/issue-54242.rs
Comment (matthewjasper) [reply src/test/incremental/issue-54242.rs:3]: So, the cycle is broken with this PR because bounds don't need to be proven to prove that the trait is well-formed. This makes the code work like it does currently.

**Comment by matthewjasper** [comment]
So this PR as is currently breaks the following:

```rust
trait Op where Self::Output: Copy {
    type Output;
}

// This now requires an explicit `where T::Output: Copy `
// like any other kind of where clause on the trait, except super traits.
fn f<T: Op>() {}
```

Since this apparently never comes up in rustc/perf/the tests I'm leaving it in for the crater run to satisfy my curiosity about whether anyone is writing bounds like this. I'll probably fix it even if crater can't find anyone doing this.

**Comment by nikomatsakis** [comment]
> So this PR as is currently breaks the following:

I was going to ask about that...thanks for reminding me. I too am curious how much it matters.

**Comment by erikdesjardins** [comment]
FWIW `mogwai` does this: https://github.com/schell/mogwai/blob/f21d6e00c4dc8780115ee9a636b4317649209ed9/mogwai/src/component.rs#L123-L128
```rust
pub trait Component
where
  Self: Sized + 'static,
  Self::ModelMsg: Clone,
  Self::ViewMsg: Clone,
  Self::DomNode: JsCast + AsRef<Node> + Clone,
```

File: src/librustc_error_codes/error_codes/E0284.md
Comment (MajorBreakfast) [inline src/librustc_error_codes/error_codes/E0284.md:17]:
```suggestion
return any type where `u32: Into<T>`.
```

**Comment by matthewjasper** [comment]
Hmm, script-servo is still broken, looks like I need more normalize calls...

**Comment by matthewjasper** [comment]
@craterbot check

File: src/librustc_infer/traits/util.rs
Comment (estebank) [reply src/librustc_infer/traits/util.rs:151]: Pretty sure I'm the culprit 👋

File: src/test/ui/ufcs/ufcs-qpath-self-mismatch.stderr
Comment (estebank) [inline src/test/ui/ufcs/ufcs-qpath-self-mismatch.stderr:8]: I'll take a look later how it can be done, but it would be better if we didn't add this redundant `note`.

File: src/test/ui/type-alias-impl-trait/bounds-are-checked.stderr
Comment (estebank) [inline src/test/ui/type-alias-impl-trait/bounds-are-checked.stderr:23]: Having type aliases for traits (namely, trait aliases) evaluating their obligations at the definition point is counter to how type aliases in general currently behave. Wouldn't we want to unify the behavior, one way or the other?

File: src/test/ui/type-alias-impl-trait/bounds-are-checked-2.stderr
Comment (estebank) [inline src/test/ui/type-alias-impl-trait/bounds-are-checked-2.stderr:10]: Re: my last comment. Oh, I see that you _are_ unifying their behavior. This will [very likely fail the crater run](https://github.com/rust-lang/rust/pull/69741#issuecomment-598973937).

File: src/librustc_trait_selection/traits/select/confirmation.rs
Comment (matthewjasper) [inline src/librustc_trait_selection/traits/select/confirmation.rs:960]: This is a bit hacky. I'm going to try making a change to how project works for trait objects for this and see if that's better,

File: src/test/ui/type-alias-impl-trait/bounds-are-checked-2.stderr
Comment (matthewjasper) [reply src/test/ui/type-alias-impl-trait/bounds-are-checked-2.stderr:10]: This is only for impl Trait.

**Comment by pietroalbini** [comment]
@craterbot abort

**Comment by pietroalbini** [comment]
@craterbot check

**Comment by matthewjasper** [comment]
"regressions" look like they're mostly bugs in this PR.

**Comment by matthewjasper** [comment]
I've fixed a everything that I think was a bug.

This needs a full crater re-run because it's possible that the fixes have introduced any unexpected bugs/regressions.
@bors try @rustc-timer queue

@nikomatsakis you might want to have another look at this before the crater run because there are a couple of significant changes since your last review.

**Comment by matthewjasper** [comment]
On the off chance that the fixes changed the performance impact @rust-timer queue

File: src/test/ui/associated-type-bounds/bounds-on-assoc-in-trait.rs
Comment (estebank) [inline src/test/ui/associated-type-bounds/bounds-on-assoc-in-trait.rs:21]: Isn't this technically a regression? (For the same reason https://github.com/rust-lang/rust/pull/69741 might not land.)

File: src/test/ui/associated-types/defaults-suitability.stderr
Comment (estebank) [inline src/test/ui/associated-types/defaults-suitability.stderr:8]: Could we change the span to only point at `NotClone`?

File: src/test/ui/associated-type-bounds/bounds-on-assoc-in-trait.rs
Comment (matthewjasper) [reply src/test/ui/associated-type-bounds/bounds-on-assoc-in-trait.rs:21]: This is an unstable feature that doesn't work correctly in this position with or without this change.

**Comment by matthewjasper** [comment]
In case I forget
@bors p=1 rollup=never

**Comment by matthewjasper** [comment]
@craterbot check

**Comment by matthewjasper** [comment]
@craterbot retry

**Comment by Mark-Simulacrum** [comment]
prioritize beta runs (will requeue in a bit): @craterbot abort

**Comment by Mark-Simulacrum** [comment]
Let's try that again...

@craterbot abort name=pr-73905-1

**Comment by Mark-Simulacrum** [comment]
@craterbot check

**Comment by matthewjasper** [comment]
This run was mostly done last I checked...
@craterbot p=1

**Comment by lqd** [comment]
it doesn't seem as bad as it looks ?

here's the E0282 from `traitgraph-0.1.0` [reduced](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a15ed50ab2973d519e910fe81bfd4525), removing either `From<Option<..>>` bound from `OptionalGraphIndex` makes it build.

**Comment by matthewjasper** [comment]
I've looked through the report now.

## Bugs in this PR

- missing normalize call
    - `stacked-0.1.3`
- fixed by #74670
    - `csm123199.aochelper.179006473113bfb2d0d576e52cb35117396e36c8`
- checking more lifetime bounds on trait objects than strictly necessary
    - `dces-0.2.0`
    - `lang_extension-1.0.1`
    - `ncollide2d-0.23.2`
    - `ncollide3d-0.23.2`
- Not deduplicating item bounds after normalization
    - `nalgebra`

## Use of non-wf trait object types

Object types like `dyn Iterator<Item = str>` were unsoundly allowed before this PR.

- `actix-amqp-0.1.4`
- `ntex-amqp-0.1.18`
- `marcianx.openmesh-rs-core.632b1987290da4936b42fd11bb85f4c1c6ac1ca2`
- `uom-0.29.0` (and all of the other versions) - 68 regressions total.
- `grdf-0.2.0-beta` (uses `feature(generic_associated_types)`)

## Use of non-wf impl trait types:

Impl trait types could not actually implement the expected trait before this PR. This only affects code that uses `feature(type_alias_impl_trait)`.

- `amadeus-serde-0.4.0`

## Ambiguous projection bounds

Given an associated type like `type A: From<u32> + From<u64>` the compiler used to arbitrarily pick one of the traits to use.

- `traitgraph-0.1.0`
- `embedded-time-0.9.0`
- `sp-arithmetic` and some other Parity crates - 50 regressions total
- `sample` - 44 regressions total

## Use of generic associated types which don't satisfy predicates on the associated type

Predicates on associated types were not being checked when projecting associated types. This only affects code that uses `feature(generic_associated_types)`.

- `map-trait-0.3.1`

So that's ~11 root regressions and ~167 total regressions.

**Comment by lqd** [comment]
@rust-timer build 4e4e6f15b872a60816b90eed0c8a72b153d4460c

**Comment by lqd** [comment]
@craterbot check crates=list:https://crater-reports.s3.amazonaws.com/pr-73905-1/retry-regressed-list.txt p=1

**Comment by lqd** [comment]
@craterbot check crates=https://crater-reports.s3.amazonaws.com/pr-73905-1/retry-regressed-list.txt p=1

**Comment by matthewjasper** [comment]
Crater run is as expected

**Comment by nikomatsakis** [comment]
@matthewjasper is anything holding this PR up from landing, apart from needing a rebase? I guess it needs a review, too?

**Comment by matthewjasper** [comment]
This just needs a review and rebase,

**Comment by tamird** [comment]
The pathological compilation-time cases addressed by this PR are blocking us over in Fuchsia. How can we help get this reviewed?

File: compiler/rustc_middle/src/ty/sty.rs
Comment (jackh726) [inline compiler/rustc_middle/src/ty/sty.rs:1516]: Can you elaborate? What's the problem here?

**Review by jackh726 — COMMENTED** [review]
I started going through this. Didn't get very far, but did have a few thoughts that I thought were worth commenting.

**Comment by casey** [comment]
I tested this PR at commit 255cb9f3bd8dd9785d0482222308680225cc0130 against a few code snippets that elicit ICEs, at the suggestion of @jackh726. I thought they might all be triggering the same ICE, but interestingly one of them stays the same between rustc nightly and the rustc I built, and two of them still produce an ICE, but with different error messages. (The error messages are much more informative, so, hey, progress!)

I created a repository with all the snippets, along with the compilation errors here:

https://github.com/casey/gat-ice
https://github.com/casey/gat-ice/blob/master/src/bin/gat-bound.rs
https://github.com/casey/gat-ice/blob/master/src/bin/issue-68648.rs
https://github.com/casey/gat-ice/blob/master/src/bin/issue-71176.rs

File: compiler/rustc_error_codes/src/error_codes/E0284.md
Comment (matthewjasper) [reply compiler/rustc_error_codes/src/error_codes/E0284.md:7]: The current example doesn't have the right error code any more.

**Comment by matthewjasper** [comment]
I'm not sure if opening another issue for gat bounds causing ICEs is really helpful. I'll get to fixing them once this has merged.

**Review by jackh726 — COMMENTED** [review]
Still going through this. A few more comments.

**Review by jackh726 — APPROVED** [review]
Just finished going through this. Looks good overall!

File: compiler/rustc_trait_selection/src/traits/select/confirmation.rs
Comment (ishitatsuyuki) [inline compiler/rustc_trait_selection/src/traits/select/confirmation.rs:373]: What does interchanging these two do? Does it affect the logic or is it just a performance optimization?

Comment (matthewjasper) [reply compiler/rustc_trait_selection/src/traits/select/confirmation.rs:373]: This affects semantics. Evaluation is done in a single pass. The change is so that evaluating `substs.obligations` resolves some type variables before evaluating `impl_obligations`.

File: compiler/rustc_infer/src/traits/util.rs
Comment (matthewjasper) [reply compiler/rustc_infer/src/traits/util.rs:100]: It looks like only the predicates from the obligations returned by `elaborate_predicates` are used.

**Comment by lzutao** [comment]
failed NLL tests: https://github.com/rust-lang-ci/rust/runs/1212475986#step:23:28977
maybe missing a bless.

**Comment by Mark-Simulacrum** [comment]
As expected, this is a major improvement in compile times (2-5% on many benchmarks, up to 80% on stress tests), see [wall-time](https://perf.rust-lang.org/compare.html?start=5849a7eca90582ee59b67eb09548a2aa424d7f52&end=08e2d4616613716362b4b49980ff303f2b9ae654&stat=wall-time). Great work!

**Comment by matthewjasper** [comment]
For whoever does the release notes this PR has two compatibility notes:

- Associated type bindings on trait objects are verified to meet the bounds declared on the trait when checking that they implement the trait. (#27675)
- When traits bounds on associated types or opaque types are ambiguous the compiler no longer makes an arbitrary choice on which bound to use. (#54121)

**Pull request merged**
